### ISSUE TITLE
Accessibility: Screen readers cut off lines in latest insider build

### ISSUE THREAD

**Issue by Menelion** [microsoft/vscode#237493]
Type: <b>Bug</b>

In this latest insiders build, screen readers (tested with both JAWS and NVDA) started cutting off lines of code. the amount read is a bit longer than what is visible on screen, but still it's nearly impossible to work because before the whole line was read no matter what.
Please fix or revert this.

VS Code version: Code - Insiders 1.97.0-insider (9b0b13d9bfe21c3dfd227bfaa8ed5693e309a2e0, 2025-01-08T05:06:32.681Z)
OS version: Windows_NT x64 10.0.26100
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|12th Gen Intel(R) Core(TM) i7-1270P (16 x 2496)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|undefined|
|Memory (System)|15.24GB (6.47GB free)|
|Process Argv|--disable-extensions . --crash-reporter-id 798ba14e-190a-4e22-8cdc-4ff2f816f42d|
|Screen Reader|yes|
|VM|0%|
</details>Extensions disabled<details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vscod805cf:30301675
vsaa593cf:30376535
py29gd2263:31024238
vscaat:30438846
c4g48928:30535728
a9j8j154:30646983
962ge761:30841072
pythonnoceb:30776497
dsvsc014:30777825
dsvsc015:30821418
pythonmypyd1:30859725
2e7ec940:31000449
pythontbext0:30879054
cppperfnew:30980852
pythonait:30973460
dvdeprecation:31040973
dwnewjupyter:31046869
newcmakeconfigv2:31071590
nativerepl1:31134653
pythonrstrctxt:31093868
nativeloc1:31118317
cf971741:31144450
parsetypescript:31116713
e80f6927:31120813
iacca1:31150324
notype1:31143044
dwcopilot:31158714
h409b430:31177054
2j25a237:31183119
c3hdf307:31184662
6074i472:31201624

```

</details>

**Comment by meganrogge** [comment]
Do you have any settings that could explain this? Nothing that I know of has changed and I cannot reproduce the issue.

**Comment by Menelion** [comment]
Nope, yesterday everything was normal, today after the update it's like this. Nothing touched, tried to disable all extensions. If the editor window is smaller, it gets even worse. Like (it's PHP code but doesn't matter): the line `use App\Entity\User;` gets read like this: `use`. that's it. to hear the rest, you have to press the End key.  
If the window is on full screen, I hear nearly 110 first characters of the line in question.

**Comment by meganrogge** [comment]
which screen reader are you using? have you tried others?

**Comment by Menelion** [comment]
JAWS and NVDA, both latest versions.

**Comment by meganrogge** [comment]
@rperez030 or @jooyoungseo, can you reproduce? I don't have my windows computer atm.

**Comment by rperez030** [comment]
@Menelion  could you please share a file where you experience this problem? So far, I cannot reproduce it either. Maybe my lines are not long enough :).

**Comment by jooyoungseo** [comment]
@meganrogge @rperez030 I have the same issue. This needs to be addressed asap. Copy and paste the following text into a new editor and test with JAWS and NVDA:

```markdown
The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog. The quick brown fox jumps over the lazy dog.
```

Previously, all the content on the same line was read aloud unless they had a hard return.

Currently, a soft line break is introduced even for screen readers and some words are cut off like this:
blank
* l1: The quick brown fox jumps over the lazy dog. The quick brown fox
* l2: dog.

There are missing words between two lines.

**Comment by rperez030** [comment]
Thanks @jooyoungseo. i can reproduce it now. @meganrogge @isidorn this is an extremely critical issue. I think this is related to #229051. We discussed this problem back then. The workaround is to add `"editor.experimentalEditContextEnabled": false ` to the user preferences file which reverse to using the textarea.

**Comment by meganrogge** [comment]
Thank you both for investigating and calling this to my attention. It makes sense that I could not reproduce as I had disabled that setting. We will ensure it's fixed promptly. @aiday-mar , while looking for a permanent fix, I suggest disabling `"editor.experimentalEditContextEnabled"` when in screen reader mode by default.

**Comment by Menelion** [comment]
Thanks folks! Do you know if this is related or not? Approximately from the same time both JAWS and NVDA started reading a non-existing right brace. I'll explain.
In PHP world it's common practice to have a blank line at the end of the file. And many files, being classes, end with a right brace `}` and a blank line. so now both JAWS and NVDA in VS Code read in speech and show in Braille the right brace on both lines, one where it does exist indeed, and another on the blank line. If you press the "read current character" key, you'll hear "blank" or "U+0" as expected.
Any ideas if this is the same issue or should I open a new one? Thanks!

**Comment by aiday-mar** [comment]
Hi @meganrogge thank you for the ping and apologies for the issue. I will disable the edit context for screen reader users and will look for a fix.

**Comment by aiday-mar** [comment]
I have the following PR which disables the edit context: https://github.com/microsoft/vscode/pull/238112. I will investigate the issue more in depth

**Comment by aiday-mar** [comment]
I am trying to reproduce this issue. I am currently not able to exactly reproduce the issue described by @jooyoungseo using NVDA on windows. I turned on word wrap and even with a line spread out on several view lines, NVDA still reads the full line. I was running however into another issue while using NVDA, which was that I had to press Escape to move the cursor up and down and be able to navigate. I am not sure if this is an edit context problem or becuase I was connecting to windows through the dev box and maybe it was an input issue.

https://github.com/user-attachments/assets/0038dcad-7d4b-46fe-b5ac-653538eba3ad

I then tried with Voice Over on macos and I noticed a different behavior where only the text on the wordwrapped line got read, not the full line (this does not happen for me on windows). I will investigate this. 

Could I ask the users in this thread for your settings, so I can try to reproduce the issue?

**Comment by Menelion** [comment]
@aiday-mar I have word wrap turned off so line numbers would match what is shown in error messages (like: "error on line 128"). How can I send you my settings please?

**Comment by aiday-mar** [comment]
Hi @Menelion thanks for your response. You can copy paste them here, or if there is confidential information if you'd prefer , I can send you an email on your profile email and you can reply with the settings

**Comment by Menelion** [comment]
@aiday-mar No, by "how" I really meant that, sorry for my stupidness. How do I copy them? Please help me locate the full JSON file you need on Windows. Thanks!

**Comment by aiday-mar** [comment]
Hi, ah yes of course, you can run the following command: `workbench.action.openSettingsJson`. In the command palette this command is called: `Preferences: Open User Settings (JSON)`. This will open a file called `settings.json` which contains all your settings.

**Comment by meganrogge** [comment]
I am able to reproduce on mac with VoiceOver when I enable`"editor.experimentalEditContextEnabled"`

**Comment by rperez030** [comment]
> I was running however into another issue while using NVDA, which was that I had to press Escape to move the cursor up and down and be able to navigate. I am not sure if this is an edit context problem or because I was connecting to windows through the dev box and maybe it was an input issue.

Make sure you are not using NVDA in browse mode. One way to ensure you are not in browse mode is trying to type some content. You can use NVDA +L (Laptop) or NVDA +Up arrow (desktop) to read the line with the cursor. you can configure your keyboard layout under NVDA menu / Preferences / Keyboard.

**Comment by rperez030** [comment]
> Approximately from the same time both JAWS and NVDA started reading a non-existing right brace. 

@Menelion  Can you please try if disabling "editor.experimentalEditContextEnabled" fix the issue with the right brace? If not, I think we should open a separate issue for that one.

**Comment by Menelion** [comment]
@rperez030 Yes yes! It fixed the non-existent line announcement issue. Actually, it's not only right brace, it doesn't matter what. To reproduce this, type something in a file and press Enter so the last line is blank, then press Up Arrow, then down Arrow again to read the last line. // CC @meganrogge @aiday-mar

**Comment by aiday-mar** [comment]
Hi everyone, thank you for commenting on this issue. I believe I have found a fix for the issue, but before merging it, I was wondering if you could spare some time to verify the issue is gone? The builds with the fix are the following:

VS Code Build: 00dda3b2bb506aee1fea3efaf6caaaab4dec106a

linux-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insider-arm64-1740645656.tar.gz
linux-armhf: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insider-armhf-1740645647.tar.gz
linux-deb-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insiders_1.98.0-1740645740_arm64.deb
linux-deb-armhf: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insiders_1.98.0-1740645731_armhf.deb
linux-deb-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insiders_1.98.0-1740646779_amd64.deb
linux-rpm-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insiders-1.98.0-1740645798.el8.aarch64.rpm
linux-rpm-armhf: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insiders-1.98.0-1740645793.el8.armv7hl.rpm
linux-rpm-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insiders-1.98.0-1740646840.el8.x86_64.rpm
linux-snap-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insider-x64-1740648978.snap
linux-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/code-insider-x64-1740645799.tar.gz
win32-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/VSCodeSetup-arm64-1.98.0-insider.exe
win32-arm64-archive: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/VSCode-win32-arm64-1.98.0-insider.zip
win32-arm64-user: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/VSCodeUserSetup-arm64-1.98.0-insider.exe
win32-x64: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/VSCodeSetup-x64-1.98.0-insider.exe
win32-x64-archive: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/VSCode-win32-x64-1.98.0-insider.zip
win32-x64-user: https://vscode.download.prss.microsoft.com/dbazure/download/insider/00dda3b2bb506aee1fea3efaf6caaaab4dec106a/VSCodeUserSetup-x64-1.98.0-insider.exe

In the build I reenabled the experimental edit context when the accessibility mode is turned on.

**Comment by aiday-mar** [comment]
We would like to get a fix in for this soon, so I have merged the PR. Please let me know if you continue seeing this in the next insiders releases.
